File: .claude/skills/PAI/Tools/extract-transcript.py
# /// script
# dependencies = [
#   "faster-whisper",
# ]
# ///
"""
extract-transcript.py

CLI tool for extracting transcripts from audio/video files using faster-whisper
Part of PAI's extracttranscript skill

Self-contained UV script with inline dependencies (PEP 723)

Usage:
    uv run extract-transcript.py <file-or-folder> [options]

Examples:
    uv run extract-transcript.py audio.m4a
    uv run extract-transcript.py video.mp4 --model large-v3 --format srt
    uv run extract-transcript.py ~/Podcasts/ --batch
"""

import sys
import os
import argparse
from pathlib import Path
from faster_whisper import WhisperModel
import json

# Supported audio/video formats
SUPPORTED_FORMATS = [
    ".m4a", ".mp3", ".wav", ".flac", ".ogg", ".aac", ".wma",
    ".mp4", ".mov", ".avi", ".mkv", ".webm", ".flv"
]

# Available models
MODELS = ["tiny", "tiny.en", "base", "base.en", "small", "small.en", "medium", "medium.en", "large-v1", "large-v2", "large-v3"]

# Output formats
OUTPUT_FORMATS = ["txt", "json", "srt", "vtt"]


def is_supported_file(file_path):
    """Check if file has supported extension"""
    return Path(file_path).suffix.lower() in SUPPORTED_FORMATS


def get_files_from_directory(dir_path):
    """Get all supported audio/video files from directory"""
    files = []
    for file_path in Path(dir_path).iterdir():
        if file_path.is_file() and is_supported_file(file_path):
            files.append(str(file_path))
    return sorted(files)


def format_time_srt(seconds):
    """Format time for SRT subtitles (HH:MM:SS,mmm)"""
    hours = int(seconds // 3600)
    minutes = int((seconds % 3600) // 60)
    secs = int(seconds % 60)
    ms = int((seconds % 1) * 1000)
    return f"{hours:02d}:{minutes:02d}:{secs:02d},{ms:03d}"


def format_time_vtt(seconds):
    """Format time for WebVTT (HH:MM:SS.mmm)"""
    hours = int(seconds // 3600)
    minutes = int((seconds % 3600) // 60)
    secs = int(seconds % 60)
    ms = int((seconds % 1) * 1000)
    return f"{hours:02d}:{minutes:02d}:{secs:02d}.{ms:03d}"


def transcribe_file(file_path, model, output_format):
    """Transcribe audio file using faster-whisper"""
    print(f"\nTranscribing: {Path(file_path).name}")
    print(f"Model: {model} | Format: {output_format}")
    print("Processing...")

    try:
        # Initialize model
        whisper_model = WhisperModel(model, device="cpu", compute_type="int8")

        # Transcribe
        segments, info = whisper_model.transcribe(file_path, beam_size=5)

        # Collect segments
        segment_list = []
        for segment in segments:
            segment_list.append({
                "start": segment.start,
                "end": segment.end,
                "text": segment.text.strip()
            })

        print(f"✓ Transcription complete ({len(segment_list)} segments)")
        return segment_list

    except Exception as e:
        raise Exception(f"Transcription failed: {str(e)}")


def format_transcript(segments, output_format):
    """Format transcript in requested format"""
    if output_format == "txt":
        return " ".join([seg["text"] for seg in segments])

    elif output_format == "json":
        return json.dumps(segments, indent=2)

    elif output_format == "srt":
        output = []
        for i, seg in enumerate(segments, 1):
            start = format_time_srt(seg["start"])
            end = format_time_srt(seg["end"])
            output.append(f"{i}\n{start} --> {end}\n{seg['text']}\n")
        return "\n".join(output)

    elif output_format == "vtt":
        output = ["WEBVTT\n"]
        for i, seg in enumerate(segments, 1):
            start = format_time_vtt(seg["start"])
            end = format_time_vtt(seg["end"])
            output.append(f"{i}\n{start} --> {end}\n{seg['text']}\n")
        return "\n".join(output)

    else:
        raise ValueError(f"Unsupported format: {output_format}")


def save_transcript(file_path, transcript, output_format, output_dir=None):
    """Save transcript to file"""
    # Determine output directory
    if output_dir:
        out_dir = Path(output_dir)
        out_dir.mkdir(parents=True, exist_ok=True)
    else:
        out_dir = Path(file_path).parent

    # Generate output filename
    base_name = Path(file_path).stem
    output_path = out_dir / f"{base_name}.{output_format}"

    # Save to file
    output_path.write_text(transcript, encoding="utf-8")

    return str(output_path)


def main():
    parser = argparse.ArgumentParser(
        description="Extract transcripts from audio/video files using faster-whisper"
    )
    parser.add_argument("path", help="File or folder path to transcribe")
    parser.add_argument(
        "--model",
        default="base.en",
        choices=MODELS,
        help="Whisper model size (default: base.en)"
    )
    parser.add_argument(
        "--format",
        default="txt",
        choices=OUTPUT_FORMATS,
        help="Output format (default: txt)"
    )
    parser.add_argument(
        "--batch",
        action="store_true",
        help="Process all files in folder"
    )
    parser.add_argument(
        "--output",
        help="Output directory (default: same as input)"
    )

    args = parser.parse_args()

    # Check if path exists
    input_path = Path(args.path).expanduser().resolve()
    if not input_path.exists():
        print(f"Error: Path does not exist: {input_path}")
        sys.exit(1)

    # Get files to process
    if input_path.is_dir():
        if not args.batch:
            print("Error: Path is a directory. Use --batch flag to process all files.")
            sys.exit(1)

        print(f"Processing directory: {input_path}")
        files = get_files_from_directory(input_path)

        if not files:
            print(f"Error: No supported audio/video files found")
            print(f"Supported formats: {', '.join(SUPPORTED_FORMATS)}")
            sys.exit(1)

        print(f"Found {len(files)} file(s) to transcribe")

    elif input_path.is_file():
        if not is_supported_file(input_path):
            print(f"Error: Unsupported file format: {input_path.suffix}")
            print(f"Supported formats: {', '.join(SUPPORTED_FORMATS)}")
            sys.exit(1)
        files = [str(input_path)]

    else:
        print(f"Error: Path is neither a file nor directory: {input_path}")
        sys.exit(1)

    # Process each file
    results = []
    errors = []

    for file_path in files:
        try:
            segments = transcribe_file(file_path, args.model, args.format)
            transcript = format_transcript(segments, args.format)
            output_path = save_transcript(file_path, transcript, args.format, args.output)
            results.append({"file": file_path, "output": output_path})
            print(f"✓ Saved to: {output_path}")
        except Exception as e:
            errors.append({"file": Path(file_path).name, "error": str(e)})
            print(f"✗ Failed to transcribe {Path(file_path).name}: {e}")

    # Summary
    print("\n" + "=" * 60)
    print("Transcription complete!")
    print(f"Successfully processed: {len(results)}/{len(files)} files")
    if errors:
        print(f"Failed: {len(errors)} files")
    print("=" * 60)

    if results:
        print("\nOutput files:")
        for result in results:
            print(f"  - {result['output']}")

    if errors:
        print("\nErrors:")
        for error in errors:
            print(f"  - {error['file']}: {error['error']}")


if __name__ == "__main__":
    main()

File: .claude/skills/PAI/Tools/ExtractTranscript.ts
#!/usr/bin/env bun

/**
 * ExtractTranscript.ts
 *
 * CLI tool for extracting transcripts from audio/video files using OpenAI Whisper API
 * Part of PAI's extracttranscript skill
 *
 * Usage:
 *   bun ExtractTranscript.ts <file-or-folder> [options]
 *
 * Examples:
 *   bun ExtractTranscript.ts audio.m4a
 *   bun ExtractTranscript.ts video.mp4 --format srt
 *   bun ExtractTranscript.ts ~/Podcasts/ --batch
 */

import OpenAI from "openai";
import { existsSync, statSync, readdirSync, mkdirSync, createReadStream } from "fs";
import { join, basename, extname, dirname } from "path";
import { writeFile } from "fs/promises";

// Supported audio/video formats
const SUPPORTED_FORMATS = [
  ".m4a",
  ".mp3",
  ".wav",
  ".flac",
  ".ogg",
  ".mp4",
  ".mpeg",
  ".mpga",
  ".webm",
];

// Output formats
const OUTPUT_FORMATS = ["txt", "json", "srt", "vtt"];

interface Options {
  format: string;
  batch: boolean;
  outputDir?: string;
}

/**
 * Parse command line arguments
 */
function parseArgs(): { path: string; options: Options } {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    console.error("Error: No file or folder path provided");
    console.log(
      "\nUsage: bun ExtractTranscript.ts <file-or-folder> [options]"
    );
    console.log("\nOptions:");
    console.log(
      "  --format <format>    Output format (txt, json, srt, vtt) [default: txt]"
    );
    console.log("  --batch              Process all files in folder");
    console.log(
      "  --output <dir>       Output directory [default: same as input]"
    );
    console.log("\nExamples:");
    console.log("  bun ExtractTranscript.ts audio.m4a");
    console.log("  bun ExtractTranscript.ts video.mp4 --format srt");
    console.log("  bun ExtractTranscript.ts ~/Podcasts/ --batch");
    console.log("\nEnvironment:");
    console.log("  OPENAI_API_KEY       Required - your OpenAI API key");
    process.exit(1);
  }

  const path = args[0];
  const options: Options = {
    format: "txt",
    batch: false,
  };

  for (let i = 1; i < args.length; i++) {
    const arg = args[i];

    if (arg === "--format" && i + 1 < args.length) {
      const format = args[i + 1];
      if (!OUTPUT_FORMATS.includes(format)) {
        console.error(
          `Error: Invalid format "${format}". Must be one of: ${OUTPUT_FORMATS.join(", ")}`
        );
        process.exit(1);
      }
      options.format = format;
      i++;
    } else if (arg === "--batch") {
      options.batch = true;
    } else if (arg === "--output" && i + 1 < args.length) {
      options.outputDir = args[i + 1];
      i++;
    }
  }

  return { path, options };
}

/**
 * Check if path is a supported audio/video file
 */
function isSupportedFile(filePath: string): boolean {
  const ext = extname(filePath).toLowerCase();
  return SUPPORTED_FORMATS.includes(ext);
}

/**
 * Get all supported files from a directory
 */
function getFilesFromDirectory(dirPath: string): string[] {
  const files: string[] = [];

  try {
    const entries = readdirSync(dirPath);

    for (const entry of entries) {
      const fullPath = join(dirPath, entry);
      const stat = statSync(fullPath);

      if (stat.isFile() && isSupportedFile(fullPath)) {
        files.push(fullPath);
      }
    }
  } catch (error) {
    console.error(`Error reading directory: ${error}`);
    process.exit(1);
  }

  return files;
}

/**
 * Get file size in MB
 */
function getFileSizeMB(filePath: string): number {
  const stats = statSync(filePath);
  return stats.size / (1024 * 1024);
}

/**
 * Transcribe audio file using OpenAI Whisper API
 * Automatically splits large files if needed
 */
async function transcribeFile(
  filePath: string,
  options: Options,
  openai: OpenAI
): Promise<string> {
  console.log(`\nTranscribing: ${basename(filePath)}`);

  const fileSizeMB = getFileSizeMB(filePath);
  console.log(`File size: ${fileSizeMB.toFixed(2)} MB`);

  // OpenAI has 25MB file size limit - use local split helper for large files
  if (fileSizeMB > 25) {
    console.log("File exceeds 25MB limit - using faster local alternative...");
    console.log("Note: For large files, consider using faster-whisper locally");
    throw new Error(
      `File size (${fileSizeMB.toFixed(2)} MB) exceeds OpenAI's 25MB limit. Please use a local transcription tool or split the file manually.`
    );
  }

  console.log(`Format: ${options.format}`);
  console.log("Uploading to OpenAI...");

  try {
    // Create file stream
    const fileStream = createReadStream(filePath) as any;

    // Call OpenAI Whisper API
    const transcription = await openai.audio.transcriptions.create({
      file: fileStream,
      model: "whisper-1",
      response_format: options.format === "txt" ? "text" : options.format as any,
      language: "en",
    });

    console.log(`✓ Transcription complete`);

    // Return as string (API returns string for all formats)
    return typeof transcription === 'string' ? transcription : JSON.stringify(transcription, null, 2);
  } catch (error: any) {
    throw new Error(`Transcription failed: ${error.message || error}`);
  }
}

/**
 * Save transcript to file
 */
async function saveTranscript(
  filePath: string,
  transcript: string,
  options: Options
): Promise<string> {
  // Determine output directory
  const outputDir = options.outputDir || dirname(filePath);

  // Create output directory if it doesn't exist
  if (!existsSync(outputDir)) {
    mkdirSync(outputDir, { recursive: true });
  }

  // Generate output filename
  const baseName = basename(filePath, extname(filePath));
  const outputPath = join(outputDir, `${baseName}.${options.format}`);

  // Save to file
  await writeFile(outputPath, transcript, "utf-8");

  return outputPath;
}

/**
 * Calculate estimated cost
 */
function calculateCost(fileSizeMB: number): string {
  // Rough estimate: 1MB ≈ 1 minute of audio
  // OpenAI charges $0.006 per minute
  const estimatedMinutes = fileSizeMB;
  const estimatedCost = estimatedMinutes * 0.006;
  return `$${estimatedCost.toFixed(3)}`;
}

/**
 * Main execution
 */
async function main() {
  // Check for API key
  if (!process.env.OPENAI_API_KEY) {
    console.error("Error: OPENAI_API_KEY environment variable not set");
    console.log("\nSet your API key:");
    console.log('  export OPENAI_API_KEY="sk-..."');
    console.log("\nOr add to ~/.zshrc for persistence:");
    console.log('  echo \'export OPENAI_API_KEY="sk-..."\' >> ~/.zshrc');
    process.exit(1);
  }

  const { path, options } = parseArgs();

  // Initialize OpenAI client
  const openai = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY,
  });

  // Check if path exists
  if (!existsSync(path)) {
    console.error(`Error: Path does not exist: ${path}`);
    process.exit(1);
  }

  // Check if it's a file or directory
  const stat = statSync(path);
  let files: string[];

  if (stat.isDirectory()) {
    if (!options.batch) {
      console.error(
        "Error: Path is a directory. Use --batch flag to process all files in folder."
      );
      process.exit(1);
    }

    console.log(`Processing directory: ${path}`);
    files = getFilesFromDirectory(path);

    if (files.length === 0) {
      console.error(`Error: No supported audio/video files found in directory`);
      console.log(`Supported formats: ${SUPPORTED_FORMATS.join(", ")}`);
      process.exit(1);
    }

    console.log(`Found ${files.length} file(s) to transcribe`);
  } else if (stat.isFile()) {
    if (!isSupportedFile(path)) {
      console.error(`Error: Unsupported file format: ${extname(path)}`);
      console.log(`Supported formats: ${SUPPORTED_FORMATS.join(", ")}`);
      process.exit(1);
    }
    files = [path];
  } else {
    console.error(`Error: Path is neither a file nor a directory: ${path}`);
    process.exit(1);
  }

  // Calculate total cost estimate
  const totalSizeMB = files.reduce((sum, file) => sum + getFileSizeMB(file), 0);
  const estimatedCost = calculateCost(totalSizeMB);

  console.log(`\nTotal size: ${totalSizeMB.toFixed(2)} MB`);
  console.log(`Estimated cost: ${estimatedCost}`);
  console.log("");

  // Process each file
  const results: Array<{ file: string; output: string }> = [];
  const errors: Array<{ file: string; error: string }> = [];

  for (const file of files) {
    try {
      const transcript = await transcribeFile(file, options, openai);
      const outputPath = await saveTranscript(file, transcript, options);
      results.push({ file, output: outputPath });
      console.log(`✓ Saved to: ${outputPath}`);
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      errors.push({ file: basename(file), error: errorMsg });
      console.error(`✗ Failed to transcribe ${basename(file)}: ${errorMsg}`);
    }
  }

  // Summary
  console.log(`\n${"=".repeat(60)}`);
  console.log(`Transcription complete!`);
  console.log(
    `Successfully processed: ${results.length}/${files.length} files`
  );
  if (errors.length > 0) {
    console.log(`Failed: ${errors.length} files`);
  }
  console.log(`${"=".repeat(60)}`);

  if (results.length > 0) {
    console.log("\nOutput files:");
    results.forEach(({ output }) => console.log(`  - ${output}`));
  }

  if (errors.length > 0) {
    console.log("\nErrors:");
    errors.forEach(({ file, error }) =>
      console.log(`  - ${file}: ${error}`)
    );
  }
}

// Run main function
main().catch((error) => {
  console.error(`Fatal error: ${error}`);
  process.exit(1);
});
